runtime/src/accounts_db.rs

```diff
@@ -6243,7 +6242,31 @@ impl AccountsDb {
         }
     }
 
+    /// sleep while accounts cache size has grown too large, up to a max wait
+    fn maybe_stall_store_cached(&self, accounts: &[(&Pubkey, &AccountSharedData)]) {
+        const MAX_DELAY_US: u128 = 10_000;
+        const CACHE_SIZE_TO_STALL: u64 = 10_000_000_000;
+        const DELAY_US: u64 = 1_000;
+        let start = Instant::now();
+        let mut waited = false;
+        while self.accounts_cache.size() > CACHE_SIZE_TO_STALL {
+            waited = true;
+            sleep(Duration::from_micros(DELAY_US));
+            if start.elapsed().as_micros() > MAX_DELAY_US {
+                break;
+            }
+        }
```

Comment:
Hmmm I'm not sure why stalling here would help if the cache was too large, and there are no other slots being replayed.

The cache should only be flushed on new roots being made, and if we stall progress here, it seems we will not complete replay of the slots, which will prevent new roots from being made, which means the cache will not be flushed.

Reply:
Cache is flushed in the bg.
stalling here helps validator keep from writing more in the fg than the bg can keep up with flushing.
There is plenty AVAILABLE to flush due to making roots, but the amount to flush keeps growing unbounded without this stall.

Reply:
I see, this is probably because the cache flush is blocked by a previous iteration of clean/shrink/hash?